### PR TITLE
Allow for GPS Data in Table/Graph/Exports

### PR DESCRIPTION
**Pull request by LinusThorsell**
PR for Issue #602 

Opening PR to discuss changes required for this feature to be completed.

### CODE REVIEW

**Comment by LinusThorsell** [comment]
Current Status:
Display data is being logged to lastGPS array.
Data is being displayed in the Table and on the Graph, BUT out of sync from the other data.

If you want to help developing this feature please post here.

**Comment by McGiverGim** [comment]
@LinusThorsell how do you know the data is out of sync of the other data? Any easy way to see that?

**Comment by LinusThorsell** [comment]
> @LinusThorsell how do you know the data is out of sync of the other data? Any easy way to see that?

Mainly assuming that it is, since the currently logged GPS Altitude data is out of sync with the new GPS Altitude data from the log I currently grab from the G frame.

image

If you have a better log file then feel free to test it and confirm or deny my assumptions!

(there is also another issue that the GPS heading value is not being properly parsed right now I think, but that is not priority 1 at the moment.)

File: js/flightlog.js
Comment (McGiverGim) [inline js/flightlog.js:462]: You are doing the same for H and G frames. If I'm not wrong, the H frame contains only two data (the coordinates), and G contains all the seven data (if GPS Home established). The loop works for both, but I suppose is better to store the fields into different variables because they are used in a different way.

File: js/flightlog.js
Comment (McGiverGim) [inline js/flightlog.js:489]: I think I understand what you are doing here, I suppose you want to "store" the latest known GPS data in the standard chunk. If I'm right, I think that what you want is to do the same that it does for the S (Slow) frame upper, in the I and P frames, look for this comment `// Then merge in the last seen slow-frame data` and do the same to add the GPS data to the standard chunk.

File: js/flightlog.js
Comment (McGiverGim) [inline js/flightlog.js:742]: If you add the data in the P and I frames, I suppose this can be removed. The data is not calculated, has been filled in the other function.

**Review by McGiverGim — COMMENTED** [review]
I didn't touch the Blackbox since a long time ago, and I don't understand totally your code but I thing you are doing one thing wrong:
- The blackbox have two kind of data. One is directly read from the log. Let's call it direct data. The other is "calculated" based on the direct data because the firmware does not write it.
- You are storing the data of the GPS, and later you are calculating it. I think this complicates the process, is better to store the data in the "standard" chunk and do nothing with the calculated fields.
I think your problem comes from here, you are not taking into account the size of the chunk in all the places, but better we can start cleaning the code with my suggestions and continue from here.

File: js/flightlog.js
Comment (LinusThorsell) [reply js/flightlog.js:462]: Thank you for the feedback and knowledgesharing!

Will look into these tonight!

File: js/flightlog.js
Comment (LinusThorsell) [reply js/flightlog.js:742]: Agree, removed it.

File: js/flightlog.js
Comment (LinusThorsell) [reply js/flightlog.js:489]: Done.

**Comment by LinusThorsell** [comment]
There we go, implemented all the suggestions you had @McGiverGim 

I think that the changes i pushed should make it work in theory, and it is almost working.

What is currently working:
GPS-Data is being merged into the frame in the same way as the slow frames.
GPS-Data is being input into the fieldName array.

What is left to figure out:
The table/graph gets the wrong values still. Look at this image:
bild
The entries in the frame that are interesting are frames 39 -> 45 as per this image:
bild
Which come from the G frame like this:
bild

But the current issue is that the values that the application reads from the frame is incorrect, for example the gps_numSat is reading the gps_ground_course frame data.

Anyone got any ideas for how this could happen and what the fix is?

**Comment by haslinghuis** [comment]
Please rebase the PR

**Comment by ctzsnooze** [comment]
@LinusThorsell this is AMAZING!  

Congratulations!

I can see that the GPS altitude being reported here is absolute altitude.  Whereas the flight code, after some recent changes, shows 'relative to home point' altitude in the OSD, and logs that value in the debugs, after the home point is set.  

So now we have to decide whether we are interested in absolute, or relative, altitude.

In most cases I think it is best to see 'relative to takeoff point' altitude.  It's a lot easier to interpret and compare to our debugs.

What do you guys think?  If we know the home point, we  can transmit absolute altitude to the log, and do the subtraction of the home altitude in the log explorer.  Or, in firmware, we can report absolute altitude until armed, then relative altitude.

**Comment by ctzsnooze** [comment]
Here is a graph showing the GPS Altitude from this PR in cyan, and the smoothed GPS altitude and Baro altitude from the debugs.  Data is from the file above.

The offset is just because we zero the value to home point in the debug.

Altitude values from this PR are PERFECT !!!! 

NB:  Having the GPS values in the log itself will make it much easier for us to set up position hold things since we can see the raw data.  

Screen Shot 2022-11-22 at 09 09 53

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> FAIL
 - cooling off period lapsed -> FAIL
 - commit count less or equal to three -> FAIL
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> FAIL
 - approver count at least three -> FAIL

**Comment by bonchan** [comment]
@LinusThorsell Ive created a PR to your branch. 
matched case keys with presenter and added the gps example graph
https://github.com/LinusThorsell/blackbox-log-viewer/pull/1

**Comment by haslinghuis** [comment]
My own version so far

[diff.txt](https://github.com/betaflight/blackbox-log-viewer/files/10061452/diff.txt)

File: js/flightlog_fields_presenter.js
Comment (ctzsnooze) [inline js/flightlog_fields_presenter.js:112]: Perhaps easier to read as:
```
'GPS_numSat': "GPS Sat Count",
```

File: js/flightlog_fields_presenter.js
Comment (ctzsnooze) [inline js/flightlog_fields_presenter.js:117]: Perhaps easier to read as:
```
'GPS_ground_course': "GPS Heading",
```

File: js/flightlog_fields_presenter.js
Comment (LinusThorsell) [reply js/flightlog_fields_presenter.js:112]: Agree, changing this.

**Comment by ctzsnooze** [comment]
After line 355, in graph_config.js, adding this will scale GPS_ground_course so that 0 degrees is at the bottom, and 360 degrees at the top, as we did for the other heading displays.

```
            } else if (fieldName == "GPS_ground_course") {
                return {
                    offset: -1800,
                    power: 1.0,
                    inputRange: 1800,
                    outputRange: 1.0
                };
```

File: js/flightlog_fields_presenter.js
Comment (LinusThorsell) [reply js/flightlog_fields_presenter.js:1000]: changed.

File: js/flightlog.js
Comment (ctzsnooze) [inline js/flightlog.js:224]: Could this approach, ie using `const name of parser`, also be used in the preceding `parser.frameDefs.S` case?

Comment (LinusThorsell) [reply js/flightlog.js:224]: Yes.

**Comment by LinusThorsell** [comment]
@McGiverGim @haslinghuis @ctzsnooze @blckmn 

I am happy with the feature as it is now.

The only issue I know of is that there are still two GPS Altitude fields, I just changed one of them to reported altitude until we have a better solution. What do you guys think of that?

Feel free to look over the changes and test the code.

I'm giving it my personal LGTM!

**Comment by ctzsnooze** [comment]
@LinusThorsell Re: "The only issue I know of is that there are still two GPS Altitude fields"....

Personally I would leave the name of the 'stock' GPS Altitude field - the one every GPS user will see 'normally' in their logs as "GPS Altitude".  Currently this is a 'raw', non-zeroed, absolute or "Above Seal Level" (ASL) GPS value, updated at whatever frequency the GPS unit sends data.

The Altitude debug also reports a value with the name "GPS Altitude".   This is the resampled GPS value from position.c.  It is 'zeroed' when the GPS system sets its `HOME` position, and is directly comparable to Baro Altitude in that debug.

It would be good if, in a separate PR, we would parse the GPS 'H' fields, even if only into the info field, since they would provide the offset value required for zeroing the ASL value.

Maybe we could re-name the 'stock' or everyday GPS Altitude value to `GPS Altitude ASL`, reflecting its absolute or 'above sea level' meaning.  

Once we have GPS 'H' field parsing, we could provide a derived field called `GPS Altitude Zeroed`, or something like that, as an alternative display element.

**Comment by ctzsnooze** [comment]
@LinusThorsell Re: "The only issue I know of is that there are still two GPS Altitude fields"....

Personally I would leave the name of the 'stock' GPS Altitude field - the one every GPS user will see 'normally' in their logs as "GPS Altitude".  Currently this is a 'raw', non-zeroed, absolute or "Above Seal Level" (ASL) GPS value, updated at whatever frequency the GPS unit sends data.

The Altitude debug also reports a value with the name "GPS Altitude".   This is the resampled GPS value from position.c.  It is 'zeroed' when the GPS system sets its `HOME` position, and is directly comparable to Baro Altitude in that debug.

It would be good if, in a separate PR, we would parse the GPS 'H' fields, even if only into the info field, since they would provide the offset value required for zeroing the ASL value.

Maybe we could re-name the 'stock' or everyday GPS Altitude value to `GPS Altitude ASL`, reflecting its absolute or 'above sea level' meaning.  

Once we have GPS 'H' field parsing, we could provide a derived field called `GPS Altitude Zeroed`, or something like that, as an alternative display element.

File: js/flightlog_fields_presenter.js
Comment (ctzsnooze) [inline js/flightlog_fields_presenter.js:994]: This is reported in cm, so should be 
```
                return `${(value/10).toFixed(1)} m`;
```

Comment (LinusThorsell) [reply js/flightlog_fields_presenter.js]: Fixed

**Review by ctzsnooze — APPROVED** [review]
Great work!  I think it's ready.
Please squash the commits.

**Comment by haslinghuis** [comment]
@ctzsnooze commits will be squashed upon merging.

**Review by haslinghuis — APPROVED** [review]

**Comment by McGiverGim** [comment]
Great work for your first PR @LinusThorsell, the next step is to draw a map with the gps path  :P

**Comment by haslinghuis** [comment]
https://cdn.discordapp.com/attachments/1033150581918924862/1044703452447637514/2022-11-22_16-55-18.mp4

**Comment by haslinghuis** [comment]
@McGiverGim https://cdn.discordapp.com/attachments/1033150581918924862/1044703452447637514/2022-11-22_16-55-18.mp4

**Pull request merged**
